[APINotes] Add a 'SwiftImportAsAccessors' entry for properties.

jrose-apple opened the pull request:
If true, Swift will import the property as a getter and setter (or just a getter for read-only properties), rather than as a Swift 'var' property. At least, it will once this lands and the Swift side gets implemented.

Part of rdar://problem/28455962

jrose-apple commented:
@DougGregor, since this is my first API notes patch, would you mind reviewing?

DougGregor commented on include/clang/APINotes/Types.h:
```diff
@@ -336,8 +336,11 @@ class VariableInfo : public CommonEntityInfo {
+
 public:
-  ObjCPropertyInfo() : VariableInfo() { }
+  ObjCPropertyInfo() : VariableInfo(), SwiftImportAsAccessorsSpecified(false) {}
```
I recommend also initializing `SwiftImportAsAccessors`, even if it seems silly.

DougGregor commented on include/clang/APINotes/Types.h:
```diff
@@ -351,6 +354,31 @@ class ObjCPropertyInfo : public VariableInfo {
+      SwiftImportAsAccessorsSpecified = true;
+      SwiftImportAsAccessors = value.getValue();
+    } else {
+      SwiftImportAsAccessorsSpecified = false;
```
Same here; let's just set `SwiftImportAsAccessors = false;`.

DougGregor commented on lib/APINotes/APINotesWriter.cpp:
```diff
@@ -606,11 +606,17 @@ namespace {
+        flags |= 1 << 0;
+        flags |= value.getValue() << 1;
+      }
+      out << flags;
```
We generally don't stream values out to the `raw_ostream`;  please use the endian writer, e.g.,

```
endian::Writer<little> writer(out);
writer.write<uint18_t>(flags);
```

jrose-apple replied:
I see it in the serialization for ObjCContextInfo, [right above](https://github.com/apple/swift-clang/blob/stable/lib/APINotes/APINotesWriter.cpp#L582-L587). Want me to change that one too?

DougGregor replied:
Oh. Huh. I guess we aren't too rigid about this one.

jrose-apple replied:
It's only because it's a single byte of flags that it's okay. I admit it's not very principled, but writing a single byte as little-endian also seems weird.

DougGregor commented on lib/APINotes/APINotesYAMLCompiler.cpp:
```diff
@@ -784,6 +786,8 @@ namespace {
         if (prop.Nullability)
           pInfo.setNullabilityAudited(*prop.Nullability);
+        if (prop.SwiftImportAsAccessors)
+          pInfo.setSwiftImportAsAccessors(*prop.SwiftImportAsAccessors);
```
This file also has a translation in the _other_ direction that you'll need to do, e.g., from the `ObjCPropertyInfo` to YAML. It's used to make sure everything round-trips through the binary format.

DougGregor reviewed (changes requested):
Please add a use of this new field in `test/APINotes/Inputs/roundtrip.apinotes` . I suspect it'll expose an issue due to missing a mapping from the 'Types' representation back to YAML.

jrose-apple committed:
[APINotes] Add a 'SwiftImportAsAccessors' entry for properties.

If true, Swift will import the property as a getter and setter (or
just a getter for read-only properties), rather than as a Swift 'var'
property. At least, it will once this lands and the Swift side gets
implemented.

Part of rdar://problem/28455962

jrose-apple commented:
Updated in response to feedback.

DougGregor reviewed (approved):
LGTM, thanks!
